runtime/src/iree/io/file_handle.c

```diff
@@ -160,6 +162,276 @@ iree_io_file_handle_flush(iree_io_file_handle_t* handle) {
   return status;
 }
 
+//===----------------------------------------------------------------------===//
+// iree_io_file_handle_t utilities
+//===----------------------------------------------------------------------===//
+
+#if IREE_FILE_IO_ENABLE
+
+// Creates a new platform file at |path| for usage as defined by |mode|.
+// The file will be extended to |initial_size| upon creation.
+// Returns IREE_STATUS_ALREADY_EXISTS if the file already exists.
+// Returns IREE_STATUS_PERMISSION_DENIED if the file cannot be created.
+// Opens an existing platform file at |path| for usage as defined by |mode|.
+// Returns IREE_STATUS_NOT_FOUND if the file does not exist.
+// Returns IREE_STATUS_PERMISSION_DENIED if the specified |mode| is disallowed.
+
+#if defined(IREE_PLATFORM_WINDOWS)
+
+static iree_status_t iree_io_file_handle_platform_open(
+    iree_io_file_mode_t mode, iree_string_view_t path, bool open_existing,
+    uint64_t initial_size,
+    iree_io_file_handle_primitive_t* out_handle_primitive) {
+  IREE_ASSERT_ARGUMENT(out_handle_primitive);
+  memset(out_handle_primitive, 0, sizeof(*out_handle_primitive));
+
+  // MAX_PATH is 260 but most systems nowadays have long paths enabled.
+  char path_str[2048] = {0};
+  iree_string_view_to_cstring(path, path_str, sizeof(path_str));
+
+  DWORD desired_access = 0;
+  if (iree_all_bits_set(mode, IREE_IO_FILE_MODE_READ)) {
+    desired_access |= GENERIC_READ;
+  }
+  if (iree_all_bits_set(mode, IREE_IO_FILE_MODE_WRITE)) {
+    desired_access |= GENERIC_WRITE;
+  }
+
+  DWORD share_mode = 0;
+  if (iree_all_bits_set(mode, IREE_IO_FILE_MODE_SHARE_READ)) {
+    share_mode |= FILE_SHARE_READ;
+  }
+  if (iree_all_bits_set(mode, IREE_IO_FILE_MODE_SHARE_WRITE)) {
+    share_mode |= FILE_SHARE_WRITE;
+  }
+
+  DWORD creation_disposition = open_existing ? OPEN_EXISTING : CREATE_ALWAYS;
+
+  DWORD flags = FILE_ATTRIBUTE_NORMAL;
+  if (iree_all_bits_set(mode, IREE_IO_FILE_MODE_RANDOM_ACCESS)) {
```

Comment:
We probably want document (and maybe validate) that FILE_MODE_RANDOM_ACCESS and FILE_MODE_SEQUENTIAL_SCAN are mutually exclusive.